rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,217 @@
+# SlateDB Range Scans
+
+<!-- TOC start (generated with https://github.com/derlin/bitdowntoc) -->
+
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [Public API](#public-api)
+- [Implementation](#implementation)
+   * [Checkpoint References](#checkpoint-references)
+   * [Memory Usage](#memory-usage)
+- [Rejected Alternatives & Follow-Ups](#rejected-alternatives-follow-ups)
+- [Updates](#updates)
+
+<!-- TOC end -->
+
+Status: Discussion
+
+Authors:
+
+* [Jason Gustafson](https://github.com/hachikuji)
+
+References:
+
+* https://github.com/slatedb/slatedb/issues/8
+* https://github.com/facebook/rocksdb/wiki/Iterator
+* https://github.com/facebook/rocksdb/wiki/Iterator-Implementation
+* https://github.com/slatedb/slatedb/pull/260
+* https://github.com/slatedb/slatedb/pull/278
+
+## Motivation
+
+
+The ability to scan the keys in a range is an essential feature for building applications.
+For example, this can be used to implement a collections of keys using a common prefix. 
+Currently, SlateDB only supports fetching individual values for known keys.
+
+## Goals
+
+The goal of this work is to provide a flexible capability to query a range of keys in the
+database. It should be possible to specify unbounded ranges to allow querying all keys,
+or bounded ranges for finer precision and efficiency.
+
+Range queries have the potential to be expensive from a resource perspective because they
+hold references to existing tables, which prevents garbage collection. Our goal here is
+to provide a simple initial approach to state management which optimizes for common cases,
+while leaving room for further optimization in the future.
+
+## Public API
+
+We introduce a new `Db::scan` API which accept a `DbRange` defining the range of the scan.
+The result will be a `DbIterator` object which iterates the current records within the range.
+Users may either provide an explicit starting range in the call to `scan`, or they
+may `seek` using the returned iterator.
+
+```rust
+impl Db {
+    /// Scan a range of keys.
+    /// 
+    /// returns a `DbIterator`
+    pub async fn scan(
```

Comment:
The db instance would have a manifest when `scan` api is called. After that, it could refresh to a new manifest either due to additional L0 writes or compaction.

Would the `DbIterator` operate on a checkpoint? I assume yes. How would we get the checkpoint that the `DbIterator` is using? 

Persisting the iterator and resuming is the use case that i am thinking of. Specific checkpoint used by iterator would be part of the metadata that we need to persist.

Reply:
Yea, I think it has to operate on a checkpoint. This is mentioned in #278 ([here](https://github.com/slatedb/slatedb/pull/278/files#diff-7475f145a0c2ca116a28fc5b1896908c2017960e950364c30fddadaee84df9caR212) and a few other places). I think we can piggyback off of the writer client's manifest update calls, so checkpoints for scans should be free. For readers, they are always operating off a checkpoint as well, and thus a scan would use the existing checkpoint.

Reply:
Makes sense. I had included a reference to the checkpoint in my prototype implementation. I'll add it here.